### DISCUSSION TITLE
Need assistance setting up pnpm for docker

### DISCUSSION THREAD

**Discussion by Arctomachine** [pnpm/pnpm#5090] [Q&A]
I am trying to deploy my project via docker. This project uses pnpm.

I am following a tutorial ([link](https://steveholgado.com/nginx-for-nextjs)) which uses npm. I changed it so it uses pnpm for my project:
```Dockerfile
FROM node:16-alpine

WORKDIR /usr/app

RUN npm i -g pnpm@7.5.0
RUN pnpm setup
RUN pnpm install -g pm2

COPY ./package.json ./
COPY ./pnpm-lock.yaml ./

RUN pnpm install

COPY ./ ./

RUN pnpm build

EXPOSE 3000

USER node

CMD ["pm2-runtime", "start", "index.js;"]
```
The problem with this code is at `RUN pnpm setup` step, which was already mentioned in #4495. With adopting solution from #4837 I now have the following code:
```Dockerfile
FROM node:16-alpine

WORKDIR /usr/app

COPY ./package.json ./
COPY ./pnpm-lock.yaml ./

RUN npm i -g npm@latest; \
    npm i -g pnpm@7.5.0; \
    npm install -g pnpm; \
    pnpm --version; \
    pnpm setup; \
    mkdir -p /usr/local/share/pnpm &&\
    export PNPM_HOME="/usr/local/share/pnpm" &&\
    export PATH="$PNPM_HOME:$PATH"; \
    pnpm bin -g &&\
    pnpm install -g pm2 &&\
    pnpm install

COPY ./ ./

RUN pnpm build

EXPOSE 3000

USER node

CMD ["pm2-runtime", "start", "index.js;"]
```
This code is immensely more difficult to both write and read, but it works at least. However, there is still a problem at `CMD` step. Console gives this error when trying to run container: `Error: Cannot find module '/usr/app/pm2-runtime'`. It seems that despite all the effort to `pnpm setup` it still tries to run global package `pm2` locally. If I do `CMD ["pnpm", "start"]` instead, everything works.
At this point I searched possible solutions, but there are none. How should I proceed with this project? I like using pnpm and would not like to return to npm/yarn.

**Comment by Planxnx** [comment]
You can try this.
It's use corepack for package-manager management and added in Node.js 16.9.0 or later
Read more: https://nodejs.org/api/corepack.html
```
FROM node:16-alpine
WORKDIR /usr/app

# Install package-manager
# use corepack for easier installation
RUN corepack enable pnpm

# Install PM2
RUN pnpm install -g pm2

# Download dependencies from lockfile
# it's make build(local) faster because if there is no change in dependencies, it will use cache, 
# but with `package.json` It's updated frequently, so often it won't cache.)
# See more: https://pnpm.io/cli/fetch
COPY ./pnpm-lock.yaml ./
RUN pnpm fetch

# Install dependencies
COPY ./package.json ./
RUN pnpm install --offline

# Build
COPY ./ ./
RUN pnpm build

# Clean up devDependencies
RUN pnpm prune --prod

EXPOSE 3000

USER node

CMD ["pm2-runtime", "start", "index.js;"]
```

and you can update pnpm by use this command
`corepack prepare pnpm@7.12.0 --activate`

**Comment by karlhorky** [comment]
I also got pnpm installed on Alpine Linux using `get-pnpm`:

- https://github.com/pnpm/pnpm/issues/784#issuecomment-1518582235
